glean-core/ios/Glean/Scheduler/MetricsPingScheduler.swift

```diff
@@ -0,0 +1,213 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/// MetricsPingScheduler facilitates scheduling the periodic assembling of metrics pings,
+/// at a given time, trying its best to handle the following cases:
+///
+/// - ping is overdue (due time already passed) for the current calendar day;
+/// - ping is soon to be sent in the current calendar day;
+/// - ping was already sent, and must be scheduled for the next calendar day.
+class MetricsPingScheduler {
+    // This struct is used for organizational purposes to keep the class constants in a single place
+    struct Constants {
+        static let logTag = "glean/MetricsPingSched"
+        static let lastMetricsPingSentDateTime = "last_metrics_ping_iso_datetime"
+        static let dueHourOfTheDay = 4
+        static let lastVersionOfAppUsed = "last_version_of_app_used"
+    }
+
+    private let logger = Logger(tag: Constants.logTag)
+
+    var timer: Timer?
+
+    /// Schedules the metrics ping collection at the due time.
+    ///
+    /// - parameters:
+    ///     * now:  A `Date` representing the current date/time
+    ///     * sendTheNextCalendarDay:  Determines whether to schedule collection for the next calendar day
+    ///                            or to attempt to schedule it for the current calendar day. If the latter and
+    ///                            we're overdue for the expected collection time, the task is scheduled for
+    ///                            immediate execution.
+    func schedulePingCollection(_ now: Date, sendTheNextCalendarDay: Bool) {
+        var fireDate = Calendar.current.date(
+            bySettingHour: Constants.dueHourOfTheDay,
+            minute: 0,
+            second: 0,
+            of: now
+        )!
+
+        // Invalidiate the timer if it's already running
+        timer?.invalidate()
+
+        // Setup timer and schedule
+        if sendTheNextCalendarDay {
+            let tomorrow = Calendar.current.date(
+                byAdding: .day,
+                value: 1,
+                to: fireDate,
+                wrappingComponents: true
+            )!
+            fireDate = tomorrow
+        }
+
+        logger.debug("Scheduling the 'metrics' ping for \(fireDate)")
+        timer = Timer(
```

Comment:
The tolerance is only a feature when you use the repeating Timer, I think.  Let me re-read the docs and I'll add a tolerance factor if it will be used.

Reply:
Yeah, we might not want to add it right away. Right now we make a promise about 4am, not 4am + random tolerance. Let's keep it that way until we learned how it performs.